src/btree/iter.rs

```diff
@@ -68,232 +94,178 @@ impl<'a> BTreeIterator<'a> {
-				self.pending_next_backend = None;
+		// Lock log over function call (no btree struct change).
+		let commit_overlay = self.commit_overlay.read();
+		let next_commit_overlay = commit_overlay.get(col as usize).and_then(|o| match direction {
+			IterDirection::Forward => o.btree_next(&self.last_key),
+			IterDirection::Backward => o.btree_prev(&self.last_key),
+		});
+		let log = self.log.read();
+		let record_id = log.last_record_id(self.col);
+		// No consistency over iteration, allows dropping lock to overlay.
+		std::mem::drop(commit_overlay);
+		if record_id != self.iter.1.record_id {
+			self.pending_backend = None;
+		}
+		let next_from_pending = self
+			.pending_backend
+			.take()
+			.and_then(|pending| (pending.direction == direction).then(|| pending.next_item));
+		let next_backend = if let Some(pending) = next_from_pending {
+			pending
+		} else {
+			match self.next_backend(record_id, self.table, &*log, direction) {
+				Ok(r) => r,
+				Err(e) => return Ok(Err(e)),
 			}
-			let next_backend = if self.pending_next_backend.is_some() &&
-				self.direction == IterDirection::Backward
-			{
-				self.pending_next_backend.take().expect("Checked above")
-			} else {
-				self.prev_backend(record_id, self.table, &*log)?
-			};
-
-			match (next_commit_overlay, next_backend) {
-				(Some((commit_key, commit_value)), Some((backend_key, backend_value))) =>
-					match (commit_key.cmp(&backend_key), commit_value) {
-						(std::cmp::Ordering::Greater, Some(value)) => {
-							self.last_key = Some(commit_key.clone());
-							self.from_seek = false;
-							self.pending_next_backend = Some(Some((backend_key, backend_value)));
-							return Ok(Some((commit_key, value)))
-						},
-						(std::cmp::Ordering::Greater, None) => {
-							self.last_key = Some(commit_key);
-							self.from_seek = false;
-							self.pending_next_backend = Some(Some((backend_key, backend_value)));
-						},
-						(std::cmp::Ordering::Less, _) => {
-							self.last_key = Some(backend_key.clone());
-							return Ok(Some((backend_key, backend_value)))
-						},
-						(std::cmp::Ordering::Equal, Some(value)) => {
-							self.last_key = Some(commit_key);
-							self.from_seek = false;
-							return Ok(Some((backend_key, value)))
-						},
-						(std::cmp::Ordering::Equal, None) => {
-							self.last_key = Some(commit_key);
-							self.from_seek = false;
-						},
+		};
+		let result = match (next_commit_overlay, next_backend) {
+			(Some((commit_key, commit_value)), Some((backend_key, backend_value))) =>
+				match (direction, commit_key.cmp(&backend_key)) {
+					(IterDirection::Backward, std::cmp::Ordering::Greater) |
+					(IterDirection::Forward, std::cmp::Ordering::Less) => {
+						self.pending_backend = Some(PendingBackend {
+							next_item: Some((backend_key, backend_value)),
+							direction,
+						});
+						if let Some(value) = commit_value {
+							Some((commit_key, value))
+						} else {
+							std::mem::drop(log);
```

Comment:
log is going to be dropped anyway just afterward (same for the other `return Err(())` places)